src/algebra/module/torsion.lean

```diff
@@ -0,0 +1,155 @@
+import ring_theory.non_zero_divisors
+
+/-!
+# Torsion submodules
+
+## Main definitions
+
+* `torsion R M a` : the `a`-torsion submodule, containing all elements `x` of `M` such that
+  `a • x = 0`.
+* `torsion'' R M S`: the `S`-torsion submodule, containing all elements `x` of `M` such that
+  `a • x = 0` for some `a` in `S`.
+* `torsion' R M` : the torsion submoule, containing all elements `x` of `M` such that  `a • x = 0`
+  for some non-zero-divisor `a` in `R`.
+
+## Main statements
+
+* `torsion' R M` is a submodule.
+* `torsion R M a` can be viewed as a `(R ⧸ R·a)`-module.
+* `no_zero_smul_divisors_iff_torsion'_bot` : a module over a domain has `no_zero_smul_divisors`
+  (that is, there is no non-zero `a`, `x` such that `a • x = 0` )iff its `torsion'` submodule is
+  trivial.
+* `quotient_torsion.torsion'_eq_bot` : quotienting by the `tosion'` submodule makes the `torsion'`
+  submodule of the new module trivial. If `R` is a domain, we can derive an instance
+  `quotient_torsion.no_zero_smul_divisors : no_zero_smul_divisors R (M ⧸ torsion' R M)`.
+
+## Notation
+
+* The notions are defined for a `comm_semiring R` and a `module R M`. Some additional hypotheses on
+  `R` and `M` are required by some lemmas.
+* The letters `a`, `b`, ... are used for scalars (in `R`), while `x`, `y`, ... are used for vectors
+  (in `M`).
+
+## Tags
+
+Torsion, submodule, module, quotient
+-/
+
+open_locale non_zero_divisors
+section defs
+variables (R M : Type*) [comm_semiring R] [add_comm_monoid M] [module R M]
+
+/-- The `a`-torsion submodule for `a` in `R`, containing all elements `x` of `M` such that
+  `a • x = 0`. -/
+def torsion (a : R) : submodule R M := (distrib_mul_action.to_linear_map _ _ a).ker
+
+/-- The `S`-torsion submodule, for `S` a multiplicative submonoid of `R`, containing all elements
+  `x` of `M` such that `a • x = 0` for some `a` in `S`. -/
+def torsion'' (S : submonoid R) : submodule R M :=
+{ carrier := { x | ∃ a : S, a • x = 0 },
+  zero_mem' := ⟨1, smul_zero _⟩,
+  add_mem' := λ x y ⟨a, hx⟩ ⟨b, hy⟩,
+    ⟨b * a,
+      by rw [smul_add, ← smul_smul, mul_comm, ← smul_smul, hx, hy, smul_zero, smul_zero, add_zero]⟩,
+  smul_mem' := λ a x ⟨b, h⟩, ⟨b, begin
+    have h' : (b : R) • x = 0 := h,
+    change (b : R) • a • x = 0, rw [smul_smul, mul_comm, ←smul_smul, h', smul_zero] end⟩ }
+
+/-- The torsion submoule, containing all elements `x` of `M` such that  `a • x = 0` for some
+  non-zero-divisor `a` in `R`. -/
+def torsion' := torsion'' R M R⁰
+
+end defs
+
+section
+variables {R M : Type*}
+
+section torsion
+variables [comm_semiring R] [add_comm_monoid M] [module R M] (a : R)
+
+@[simp] lemma smul_torsion (x : torsion R M a) : a • x = 0 := subtype.ext x.prop
+@[simp] lemma smul'_torsion (x : torsion R M a) : a • (x:M) = 0 := x.prop
+
+end torsion
+
+section quotient
+variables [comm_ring R] [add_comm_group M] [module R M] (a : R)
+open ideal.quotient
+
+instance : has_scalar (R ⧸ ideal.span ({a} : set R)) (torsion R M a) :=
```

Comment:
We have a notation for this situation:
```suggestion
instance : has_scalar (R ⧸ (R ∙ a)) (torsion R M a) :=
```
Note that the `∙` character is not the usual `•` (even though its glyph is nearly identical in most fonts).

Reply:
oh... actually I tried using this notation but I ended up being misled by this note (at `linear_algebra.basic`, line 1065) : 
```/- This is the character `∙`, with escape sequence `\.`, and is thus different from the scalar
multiplication character `•`, with escape sequence `\bub`. -/
notation R`∙`:1000 x := span R (@singleton _ _ set.has_singleton x)```

Actually, it turns out typing `\.` on VSCode doesn't give me this character `∙` but another `·` (which isn't smul either). Copy-pasting your `∙` indeed makes the `R ∙ a` notation work, while typing `\.` doesn't.